### PR TITLE
Users/rohit/spn cleanup

### PR DESCRIPTION
**Pull request by RohitYadav-msft**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.Framework/Helpers/RemoteApiHelper.ps1 [hunk 1]
```diff
@@ -13,7 +13,9 @@ class RemoteApiHelper {
     {
             try {
                 $accessToken = [RemoteApiHelper]::GetAccessToken()
-                $result = Invoke-WebRequest -Uri $([RemoteApiHelper]::ApiBaseEndpoint + $uri) `
+                $uri = ([RemoteApiHelper]::ApiBaseEndpoint + $uri) #"http://localhost:5000/api" + $uri #
+
```
Comment (SINIKI) [inline src/AzSK.Framework/Helpers/RemoteApiHelper.ps1:17]: Please undo this change. Keep the original changes

File: src/AzSK.Framework/Helpers/RemoteApiHelper.ps1 [hunk 2]
```diff
@@ -107,6 +108,10 @@ class RemoteApiHelper {
     static [PSObject] GetComplianceSnapshot([string] $parameters){
 		return([RemoteApiHelper]::GetJsonContent("/compliancedata", $parameters) )	
     }
+
+    static [PSObject] GetSPNList([PSObject] $ownedSPNDetails) {
+        return([RemoteApiHelper]::PostJsonContent("/scanresults/SPNList", $ownedSPNDetails))
```
Comment (SINIKI) [inline src/AzSK.Framework/Helpers/RemoteApiHelper.ps1:113]: Should we change this API: /scanresults/SPNList

Comment (SINIKI) [reply src/AzSK.Framework/Helpers/RemoteApiHelper.ps1:113]: We can rename this function from 'GetSPNList'  to FetchUsedSPNList

File: src/AzSK.Framework/Helpers/WebRequestHelper.ps1 [hunk 3]
```diff
@@ -80,6 +80,11 @@ class WebRequestHelper {
 		return "https://management.core.windows.net/"
 	}
 
+	hidden static [string] GetUserSPNsUrl()
```
Comment (SINIKI) [inline src/AzSK.Framework/Helpers/WebRequestHelper.ps1:83]: Rename this function GetUserSPNsUrl to GetGraphAPIURL

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 4]
```diff
@@ -0,0 +1,133 @@
+using namespace System.Management.Automation
+Set-StrictMode -Version Latest 
+
+class SPNInfo: CommandBase
+{    
+	hidden $validatedUri = [string]::Empty
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:6]: We can check when is this 'validatedUri' variable used in DevOps kit

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			foreach ($item in $SPNApplicationId) 
+			{
+				$usedSPNs += $item;
+			}
+		}
+		return ($usedSPNs);
+	}
+
+
+	[PSObject] GetOwnedSPNStartWithAzSK($ownedSPNDetails)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:100]: GetOwnedSPNStartWithAzSK -> FilterOwnedSPNStartWithAzSK

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	GetSPNInfo()
+	{
+        $this.PublishCustomMessage([Constants]::DoubleDashLine + "`r`nFetching SPNs Details...`r`n" + [Constants]::DoubleDashLine);
+		$this.PublishCustomMessage("`r`nSPNs Info`r`n" + [Constants]::SingleDashLine, [MessageType]::Default)
+		#Get all owned SPNs
+        $ownedSPNDetails = $this.GetOwnedSPNList();
+		#Get SPNs start with AzSK_CA
+        $ownedSPNs = $this.GetOwnedSPNStartWithAzSK($ownedSPNDetails);
+		
+        if($ownedSPNs.count -ne 0 -and $ownedSPNs -ne $null)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:23]: change this condition:

$ownedSPNs -ne $null -and $ownedSPNs.count -ne 0

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        {
+			#Get Http response which contain used SPNs applicationId
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:25]: Update comment and function name

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            $SPNs = [RemoteApiHelper]::GetSPNList($ownedSPNs.appId);
+		    #Convert HttpResponse from json and Get CAUsedSPNs
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:27]: Rename $SPNs variable. Add 'Reponse' in the variable name

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+				return($null);
+			}
+        }
+        catch 
+        {
+            throw $_
+        }
+	} 
+
+	[PSObject] GetUsedSPNs($SPNApplicationId)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:85]: rename variable->SPNApplicationId

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+			#Get Http response which contain used SPNs applicationId
+            $SPNs = [RemoteApiHelper]::GetSPNList($ownedSPNs.appId);
+		    #Convert HttpResponse from json and Get CAUsedSPNs 
+		    $usedSPNs = $this.GetUsedSPNs($SPNs);
+		    #Get Not in CA Used SPNs
+		    $notInUsedSPNs = $this.GetUnusedSPNs($usedSPNs,$ownedSPNs);
+        
+            $this.PublishCustomMessage("`r`nIn CA used SPNs`r`n`n" + [Constants]::SingleDashLine, [MessageType]::Default)
+                
+            if($usedSPNs.count -le 1)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:34]: Check usedSPNs count with Measure-Object

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        if($ownedSPNs.count -ne 0 -and $ownedSPNs -ne $null)
+        {
+			#Get Http response which contain used SPNs applicationId
+            $SPNs = [RemoteApiHelper]::GetSPNList($ownedSPNs.appId);
+		    #Convert HttpResponse from json and Get CAUsedSPNs 
+		    $usedSPNs = $this.GetUsedSPNs($SPNs);
+		    #Get Not in CA Used SPNs
+		    $notInUsedSPNs = $this.GetUnusedSPNs($usedSPNs,$ownedSPNs);
+        
+            $this.PublishCustomMessage("`r`nIn CA used SPNs`r`n`n" + [Constants]::SingleDashLine, [MessageType]::Default)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:32]: Change this print message: In CA used SPNs

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+                $this.PublishCustomMessage("Count : 0", [MessageType]::Default);
+            }
+            else
+            {
+                $this.PublishCustomMessage($($usedSPNs | Format-Table @{Label = "SPNApplicationId"; Expression = { $_.appId } },@{Label = "SubscriptionId"; Expression = { $_.subscriptionId }} -AutoSize -Wrap | Out-String), [MessageType]::Default)
+            }
+        
+            $this.PublishCustomMessage([Constants]::SingleDashLine, [MessageType]::Default);
+        
+            $this.PublishCustomMessage("`r`nNot in CA used SPNs`r`n`n" + [Constants]::SingleDashLine, [MessageType]::Default)
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:45]: Change this print message: Not in CA used SPNs

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+		try
+        {
+			$ResourceAppIdURI = [WebRequestHelper]::GetUserSPNsUrl()
+			$accessToken = Get-AzSKAccessToken -ResourceAppIdURI $ResourceAppIdURI
+			
+			$header = "Bearer " + $accessToken
+			$headers = @{"Authorization"=$header;"Content-Type"="application/json";}
+
+		    $uri = $this.validatedUri;
+		    $result = invoke-webrequest -Uri $uri -Headers $headers -method Get -UseBasicParsing
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:68]: We can use WebRequestHelper function for this

Comment (SINIKI) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:68]: Next link not handled

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+			if($null -ne $result -and $result.StatusCode -eq 200)
+            {
+                $result_content = $result.Content | convertfrom-json;
+    		    $ownedSPNDetails = $result_content.value | Select-Object -Property displayName,appId -Unique 
+    		    return ($ownedSPNDetails);
+			}
+			else {
+				return($null);
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:76]: Why are we returning null

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+			}
+        }
+        catch 
+        {
+            throw $_
```
Comment (SINIKI) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:81]: Make sure if you are throwing error, it is being handled

File: src/AzSK.Framework/Helpers/WebRequestHelper.ps1 (hunk 3 above)
Comment (RohitYadav-msft) [reply src/AzSK.Framework/Helpers/WebRequestHelper.ps1:83]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 4 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:6]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 5 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:100]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 6 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:23]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 8 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:27]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 9 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:85]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 10 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:34]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 11 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:32]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 12 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:45]: Done

File: src/AzSK.Framework/Helpers/RemoteApiHelper.ps1 (hunk 2 above)
Comment (RohitYadav-msft) [reply src/AzSK.Framework/Helpers/RemoteApiHelper.ps1:113]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 7 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:25]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 13 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:68]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 14 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:76]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 15 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:81]: Done

File: src/AzSK.Framework/Helpers/RemoteApiHelper.ps1 (hunk 1 above)
Comment (RohitYadav-msft) [reply src/AzSK.Framework/Helpers/RemoteApiHelper.ps1]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 16]
```diff
@@ -0,0 +1,148 @@
+using namespace System.Management.Automation
+Set-StrictMode -Version Latest 
+
+class SPNInfo: CommandBase
+{    
+	hidden $validatedUri = [string]::Empty
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:6]: Give better name instead of validated Uri.. It can be like GraphOwnedObjectsAPIUri?

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    { 
+		$this.DoNotOpenOutputFolder = $true;
+		$this.validatedUri = 'https://graph.microsoft.com/v1.0/me/ownedObjects';
+	}
+	
+	GetSPNInfo()
+	{
+        $rmContext = [ContextHelper]::GetCurrentRMContext();
+        $this.PublishCustomMessage([Constants]::DoubleDashLine + "`r`nFetching SPN(s) Details...`r`n" + [Constants]::DoubleDashLine);
+		#$this.PublishCustomMessage("`r`nSPNs Info`r`n" + [Constants]::SingleDashLine, [MessageType]::Default)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:18]: Remove commented code

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+		return ($usedSPNs);
+	}
+
+
+	[PSObject] FilterOwnedSPNStartWithAzSK($ownedSPNDetails)
+	{
+		$ownedSPNUsedInCA = @();
+		if($null -ne  $ownedSPNDetails)
+		{
+			foreach ($ownedSPN in $ownedSPNDetails)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:110]: Do we really need foreach here. Can we use simply where clause itself

Comment (VishalHaibatpure) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:110]: This function itself can be replaced with single line

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+		#$this.PublishCustomMessage("`r`nSPNs Info`r`n" + [Constants]::SingleDashLine, [MessageType]::Default)
+		#Get all owned SPNs
+        $ownedSPNDetails = $this.GetOwnedSPNList();
+		#Get SPNs start with AzSK_CA
+        $ownedSPNs = $this.FilterOwnedSPNStartWithAzSK($ownedSPNDetails);
+		
+        if($ownedSPNs -ne $null -and ($ownedSPNs | Measure-Object).Count -ne 0)
+        {
+			#Add subscriptionName and subscriptionId in ownedSPNs list
+			$ownedSPNsObject = $this.CreateOwnedSPNObject($ownedSPNs.appId,$rmContext)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:27]: Are we mapping all SPN's against single subscription (current context sub)? SPN's may have used in other subs. How can we confirm all SPNs are from this sub only?

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+			
+        }
+        catch 
+        {	
+			#Exception get handle in base class
+            throw $_
+        }
+	} 
+
+	[PSObject] GetUsedSPNs($SPNsResponseObject)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:90]: Instead of full function this should have been single line where clause

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+				{
+					$ownedSPNUsedInCA += $ownedSPN;
+				}
+			}
+		}
+		return ($ownedSPNUsedInCA);
+	} 
+
+	
+	[PSObject] GetUnusedSPNs($usedSPNs,$ownedSPNs)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:122]: same as above

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+			}
+		}
+		return ($notInUsedSPNs);
+	}
+
+	[PSObject] CreateOwnedSPNObject($ownedSPNApplicationId,$rmcontext)
+	{
+		$ownedSPNObject = "" | Select-Object subscriptionId, subscriptionName, userSPNAppId
+		$ownedSPNObject.subscriptionId = $rmcontext.Subscription.Id
+		$ownedSPNObject.subscriptionName = $rmcontext.Subscription.Name
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:143]: we are mapping every SPN with same subscription. see comment above

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+	} 
+
+	
+	[PSObject] GetUnusedSPNs($usedSPNs,$ownedSPNs)
+	{
+		$notInUsedSPNs = @();
+		if($null -ne $ownedSPNs)
+		{
+			foreach ($SPNId in $ownedSPNs) 
+			{
```
Comment (TarunKrShukla) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:128]: Plz do test if there is only one SPN $ownedSPN should not be treated as single object

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 16 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:6]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 17 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:18]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 18 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:110]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 20 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:90]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 21 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:122]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 23 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:128]: Done

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 16 above)
Comment (VishalHaibatpure) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:6]: I still see same name validatedUri ?

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 22 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1]: Have removed SPN subName and subId

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 16 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1]: I have removed this with suggested variable name->GraphOwnedObjectsAPIUri

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 (hunk 19 above)
Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1]: We are not getting detail of SPN in which sub it is using, so removed context info[subId, subName] from SPN info

File: src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1 [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+			
+            $this.PublishCustomMessage("`r`nSPN(s) owned by you which are currently being used by CA :`r`n",[MessageType]::Default) # + [Constants]::SingleDashLine, [MessageType]::Default
+                
+            if(($usedSPNs | Measure-Object).Count -eq 0)
+            {
+                $this.PublishCustomMessage("`r`nCurrently there is no SPN owned by you, which is being used by CA.", [MessageType]::Warning);
+            }
+            else
+            {
+                $this.PublishCustomMessage($($usedSPNs | Format-Table @{Label = "ApplicationId"; Expression = { $_.appId } },@{Label = "SubscriptionId"; Expression = { $_.subscriptionId }},@{Label = "SubscriptionName"; Expression = { $_.subscriptionName } } -AutoSize -Wrap | Out-String), [MessageType]::Default)
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1:56]: Print application name.

Comment (RohitYadav-msft) [reply src/AzSK/Framework/Core/AzSKInfo/SPNInfo.ps1]: Done
